src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs

```diff
@@ -0,0 +1,256 @@
+                    .GetUtcNow()
+                    .AddSeconds(expiresIn);
+
+                tokensToStore.Add(new AuthenticationToken
+                {
+                    Name = "expires_at",
+                    Value = expiresAt.ToString("o", CultureInfo.InvariantCulture)
+                });
+            }
+
+            properties.StoreTokens(tokensToStore);
+        }
+
+        var identity = new ClaimsIdentity(ClaimsIssuer);
+        var ticket = await CreateTicketAsync(identity, properties, tokens);
+        return HandleRequestResult.Success(ticket);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+    {
+        if (!context.Properties.Items.TryGetValue(VkIdAuthenticationConstants.AuthenticationProperties.DeviceId, out var deviceId) ||
+            string.IsNullOrEmpty(deviceId))
+        {
+            return OAuthTokenResponse.Failed(new Exception("Device ID was not found."));
+        }
+
+        if (!context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier) ||
+            string.IsNullOrEmpty(codeVerifier))
+        {
+            return OAuthTokenResponse.Failed(new Exception("Code verifier key was not found."));
+        }
+
+        context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        var query = new Dictionary<string, string>()
+        {
+            ["grant_type"] = "authorization_code",
+            ["code"] = context.Code,
+            ["code_verifier"] = codeVerifier,
+            ["client_id"] = Options.ClientId,
+            ["device_id"] = deviceId,
+            ["redirect_uri"] = context.RedirectUri,
+            ["state"] = Options.StateDataFormat.Protect(context.Properties),
+        };
+
+        using var request = new HttpRequestMessage(HttpMethod.Post, Options.TokenEndpoint);
+        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        request.Content = new FormUrlEncodedContent(query);
+        request.Version = Backchannel.DefaultRequestVersion;
+
+        var response = await Backchannel.SendAsync(request, Context.RequestAborted);
+        var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
+        if (!payload.RootElement.TryGetProperty("state", out var state) ||
+            Options.StateDataFormat.Unprotect(state.GetString()) is null)
+        {
+            return OAuthTokenResponse.Failed(new Exception("The oauth state was missing or invalid."));
+        }
+
+        if (!payload.RootElement.TryGetProperty("error", out var errorElement))
+        {
+            return OAuthTokenResponse.Success(payload);
+        }
+
+        var errorCode = errorElement.GetString()!;
+        var errorDescription = errorElement.GetProperty("error_description").GetString()!;
+        return OAuthTokenResponse.Failed(new Exception($"{errorCode}: {errorDescription}"));
```

Comment:
Can you factor this to match the pattern used in the other providers similar to this please? For example, the error and success cases are other way around and the HTTP status code isn't being explicitly checked, so a JSON response without an `error` property is always treated as success, even if it returns a non-success response.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/82d20e3a415be2203257ecc1222edde5766961bb/src/AspNet.Security.OAuth.Deezer/DeezerAuthenticationHandler.cs#L47-L59

Similarly for other methods in this handler class in terms of layout and style - that makes the providers easier to maintain if we have to make a similar change in a future version to all 90+ providers.

Not being similar is partly why I didn't realise this already handled PKCE at a glance, because it was structured differently and the accompanying comment like you find in the other providers.

Reply:
Got it! gonna refactor it later this week.
Also, should I add logger to this handler? If yes, are there any docs about event ids?

Reply:
Just whatever logging makes sense in the same style as the other providers. The event IDs are just arbitrary numbers starting from 1 as the provider is implemented.